File: src/java/com/distributedlife/mahjong/reference/permute/AnyPairedPermutator.java
package com.distributedlife.mahjong.reference.permute;

import com.distributedlife.mahjong.reference.hand.HandCandidate;

import java.util.HashSet;
import java.util.Set;

public class AnyPairedPermutator extends Permutator {
    @Override
    public Set<HandCandidate> permute(Set<HandCandidate> candidates) {
        Set<HandCandidate> handCandidates = new HashSet<HandCandidate>();

        for (HandCandidate candidate : candidates) {
            for (String tile : candidate.getRequiredTiles()) {
                if (candidate.canAdd(tile)) {
                    HandCandidate variation = candidate.fork();

                    variation.add(tile);
                    handCandidates.add(variation);
                }
            }
        }

        return handCandidates;
    }
}

File: lib/permute/any-paired-permutator.js
/**
 * AnyPairedPermutator - Adds another of any tile already in the hand
 */
import { Permutator } from './permutator.js';

export class AnyPairedPermutator extends Permutator {
  permute(candidates) {
    const handCandidates = new Set();

    for (const candidate of candidates) {
      for (const tile of candidate.getRequiredTiles()) {
        if (candidate.canAdd(tile)) {
          const variation = candidate.fork();
          variation.add(tile);
          handCandidates.add(variation);
        }
      }
    }

    return handCandidates;
  }
}
